File: sandbox/demo2/HttpUser.java
import io.github.wfouche.tulip.api.*;

import static io.restassured.RestAssured.*;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

public class HttpUser extends TulipUser {

    public HttpUser(int userId, int threadId) {
        super(userId, threadId);
    }

    public boolean onStart() {
        // Initialize RestAssured only once
        if (getUserId() == 0) {
            baseURI = baseURI = getUserParamValue("baseURI");
        }
        return true;
    }

    // Action 1: delay 10ms
    public boolean action1() {
        try { Thread.sleep(10); } catch (InterruptedException e) { e.printStackTrace(); }
        return true;
    }

    // Action 2: delay 20ms
    public boolean action2() {
        try { Thread.sleep(20); } catch (InterruptedException e) { e.printStackTrace(); }
        return true;
    }

    // Action 3: No-op
    public boolean action3() {
        return true;
    }

    // Action 4: GET /posts/{id} using Rest-Assured
    public boolean action4() {
        boolean rc = true;
        try {
            given()
            .when()
                .get("/posts/" + Integer.toString(getUserId()+1))
            .then()
                .statusCode(200);
        } catch (java.lang.AssertionError e) {
           rc = false;
        }
        return rc;
    }

    // Action 5: GET /posts/{id} using java.net.http.HttpClient
    public boolean action5() {
        boolean rc = true;
        try {
            var response = client.send(httpRequestPosts, HttpResponse.BodyHandlers.ofString());
            rc =  (response.statusCode() == 200);
        } catch (java.lang.Exception e) {
            rc = false;
        }
        return rc;
    }

    public boolean onStop() {
        return true;
    }

    // Action 5 support data and methods
    private static HttpClient client = HttpClient.newHttpClient();

    private HttpRequest httpRequestPosts = createHttpRequest("posts");

    private HttpRequest createHttpRequest(String name) {
        try {
            var id = getUserId() + 1;
            var url = this.getUserParamValue("baseURI");
            var request = HttpRequest.newBuilder()
                    .uri(new URI(url + "/" + name + "/" + String.valueOf(id)))
                    .GET()
                    .build();
            return request;
        } catch (java.lang.Exception e) {
            throw new RuntimeException(e);
        }

    }

}

File: tulip-standalone/mbench-gradle-amper/src/HttpUser.kt
/*-------------------------------------------------------------------------*/

import io.github.wfouche.tulip.core.Console
import io.github.wfouche.tulip.api.TulipUtils
import io.github.wfouche.tulip.api.TulipUser
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse

/*-------------------------------------------------------------------------*/

private val client = HttpClient.newHttpClient()

private fun serviceCall(request:HttpRequest): Boolean {
    // https://www.baeldung.com/java-httpclient-connection-management

    val response = client.send(request, HttpResponse.BodyHandlers.ofString())

    //println(id)
    //println(name)
    //println(response.statusCode())
    //println(response.body())

    return (response.statusCode() == 200)
}

/*-------------------------------------------------------------------------*/

class HttpUser(userId: Int, threadId: Int) : TulipUser(userId, threadId) {

    // ----------------------------------------------------------------- //

    private val requestPosts = createRequest("posts")
    private val requestComments = createRequest("comments")
    private val requestAlbums = createRequest("albums")
    private val requestPhotos = createRequest("photos")
    private val requestTodos = createRequest("todos")

    override fun onStart(): Boolean {
        val actionId = 0
        Console.put("  $userId -> $actionId -> ${getActionName(actionId)}")
        return true
    }

    // ----------------------------------------------------------------- //

    override fun action1(): Boolean {
        // 6 ms delay (average)
        TulipUtils.delayMillisRandom(0, 12)
        return true
    }

    override fun action2(): Boolean {
        // 14 ms delay (average)
        TulipUtils.delayMillisRandom(0, 28)
        return true
    }

    // ----------------------------------------------------------------- //

    override fun action3(): Boolean {
        return serviceCall(requestPosts)
    }

    override fun action4(): Boolean {
        return serviceCall(requestComments)
    }

    override fun action5(): Boolean {
        return serviceCall(requestAlbums)
    }

    override fun action6(): Boolean {
        return serviceCall(requestPhotos)
    }

    override fun action7(): Boolean {
        return serviceCall(requestTodos)
    }

    // ----------------------------------------------------------------- //

    override fun action8(): Boolean {
        val actionId = 8
        Console.put("  $userId -> $actionId -> ${getActionName(actionId)}")
        return true
    }

    // ----------------------------------------------------------------- //

    override fun action9(): Boolean {
        return true
    }

    // ----------------------------------------------------------------- //

    override fun action10(): Boolean {
        Thread.sleep(10)
        return true
    }

    // ----------------------------------------------------------------- //

    override fun onStop(): Boolean {
        Console.put("  Terminate: UserId = $userId")
        Thread.sleep(100)
        return true
    }

    // ----------------------------------------------------------------- //

    private fun createRequest(name: String): HttpRequest {
        val id = userId + 1
        val url: String = this.getUserParamValue("url")
        val request:HttpRequest = HttpRequest.newBuilder()
            .uri(URI("${url}/${name}/${id}"))
            .GET()
            .build()
        return request
    }

    // ----------------------------------------------------------------- //

    companion object {
        init {
            //Console.put("Loading .... Kotlin class ... HttpUser")
        }
    }

    // ----------------------------------------------------------------- //

}

/*-------------------------------------------------------------------------*/
